src/Internal/Payment/State/InitialState.php

```diff
@@ -169,4 +192,52 @@ protected function populate_context_from_order() {
 		);
 		$context->set_customer_id( $customer_id );
 	}
+
+	/**
+	 * Detects duplicate orders, and run the necessary actions if one is detected.
+	 *
+	 * @return AbstractPaymentState|null The next state, or null if no duplicate order is detected.
+	 * @throws Order_Not_Found_Exception
+	 * @throws StateTransitionException
+	 * @throws ContainerException        When the dependency container cannot instantiate the state.
+	 */
+	protected function process_duplicate_order(): ?AbstractPaymentState {
+		$context          = $this->get_context();
+		$current_order_id = $context->get_order_id();
+
+		$duplicate_order_id = $this->dpps->get_previous_paid_duplicate_order_id( $current_order_id );
+		if ( null === $duplicate_order_id ) {
+			$this->dpps->update_session_processing_order( $current_order_id );
+			return null;
+		}
+
+		$this->dpps->clean_up_when_detecting_duplicate_order( $duplicate_order_id, $current_order_id );
+		$context->set_duplicate_order_id( $duplicate_order_id );
+		return $this->create_state( DuplicateOrderDetectedState::class );
+	}
+
+	/**
+	 * Detects duplicate payment, and run the necessary actions if one is detected.
+	 *
+	 * @return AbstractPaymentState|null The next state, or null if duplicate payment is detected.
+	 * @throws Order_Not_Found_Exception
+	 * @throws StateTransitionException
+	 * @throws ContainerException        When the dependency container cannot instantiate the state.
+	 */
+	protected function process_duplicate_payment(): ?AbstractPaymentState {
+		$context  = $this->get_context();
+		$order_id = $context->get_order_id();
+
+		$authorized_intent = $this->dpps->get_authorized_payment_intent_attached_to_order( $order_id );
+		if ( null === $authorized_intent ) {
+			return null;
+		}
+
+		$this->dpps->remove_session_processing_order( $order_id );
+		$context->set_intent( $authorized_intent );
+		$context->set_detected_authorized_intent();
+
+		$new_state = $this->create_state( ProcessedState::class );
+		return $new_state->complete_processing();
```

Comment:
@RadoslavGeorgiev - I am moving your question to here so it's not getting lost. 

> Question: Should we call `remove_session_processing_order` when transitioning from `ProcessedState` to `CompletedState`?

Yes, before we have already set $order_id when no duplicate order is detected.
https://github.com/Automattic/woocommerce-payments/blob/963541d30f6e78653c57325e52712fc417bdf130/src/Internal/Payment/State/InitialState.php#L210

This is also a bit more defensive even though the `clear_session_processing_order_after_landing_order_received_page` hook would also clear it when it arrives the order-received page.

Reply:
Sorry @htdat, could you please re-phrase this reply?  Overall, I'm not sure whether you agree we should do it, or not :D

> Yes, before we have already set $order_id when no duplicate order is detected.

Is that the `$order_id` in the session? Because we would overwrite it shortly.

Reply:
@RadoslavGeorgiev - Aha, sorry. My previous reply meant "yes, we should do that" and actually, we're doing that around this code (line 236). It's just that we're not doing that exactly when transitioning from `ProcessedState` to `CompletedState`? Instead, it's in the process of transiting from `InitialState` to `ProcessedState` (line 236 here). 

> Is that the `$order_id` in the session? Because we would overwrite it shortly.

Short answer is yes. 

So the process looks like: 

- Run `process_duplicate_order`. If no duplicate order is detected, return `null` as well as call `update_session_processing_order`. Code ref: 
https://github.com/Automattic/woocommerce-payments/blob/e688822fcac55cfd9bb6e66b947c376c092d355b/src/Internal/Payment/State/InitialState.php#L209-L213

- Run `process_duplicate_payment`. If the duplicate payment is detected, remove the `$order_id`, which has just been set.

Reply:
Ok, I think there's a gap here. Let me explain my understanding:

- The payment process starts, and in the normal flow (there is no duplicate order detected), we store the order ID in the session.
- If processing fails, we keep the order in session, and use it to prevent duplicates in the future.
- __If processing succeeds, we remove the order ID from the session.__

The last point is what we do on the Order Received page, but we should also do it at the end of the successful process, like presently do in the gateway: https://github.com/Automattic/woocommerce-payments/blob/ea5370017345fc2254f5011e4fdf381b3842e18a/includes/class-wc-payment-gateway-wcpay.php#L1501

This can only happen _after_ successful processing, but not before it, right? So in addition to what you described (`InitialState` -> `ProcessedState`), we should also clear the order from session when transition from `ProcessedState` to `CompletedState`. Am I missing something?

Reply:
That's correct 💯 . It was me to miss this fact when merging Zvonimir's PR that introduces the `ProcessedState`. Addressed here c7a932d0f40be7d205d4c06651b794cef06d58f0